File: sdk/typescript/src/services/metrics.ts
import { ClientBase } from "../base";
import type { MetricDashboardCreate, MetricDashboardUpdate } from "../models/inputs";
import type { MetricDashboardSchema, MetricsLimits, MetricsResponse } from "../models/outputs";
import type { ProductBillingType, TimeInterval, Timezone } from "../models/literals";
import { HTTPValidationError } from "../errors";

export const getMetrics = (client: ClientBase) => {
  /**
   * Get metrics about your orders and subscriptions.
   *
   * Currency values are output in cents.
   *
   * **Scopes**: `metrics:read`
   *
   * @param query - Query parameters
   * @returns {MetricsResponse}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (query: {
    start_date: string;
    end_date: string;
    timezone?: Timezone;
    interval: TimeInterval;
    organization_id?: string | string[] | null;
    product_id?: string | string[] | null;
    billing_type?: ProductBillingType | ProductBillingType[] | null;
    customer_id?: string | string[] | null;
    metrics?: string[] | null;
  }): Promise<MetricsResponse> => {
    const pathParams = {};
    const queryParams = {
      start_date: query.start_date,
      end_date: query.end_date,
      timezone: query?.timezone || "UTC",
      interval: query.interval,
      organization_id: query?.organization_id,
      product_id: query?.product_id,
      billing_type: query?.billing_type,
      customer_id: query?.customer_id,
      metrics: query?.metrics,
    };
    const request = client.buildRequest("GET", "/v1/metrics/", pathParams, queryParams, undefined);
    const response = await client.sendRequest(request);
    return client.parseResponse<MetricsResponse>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
export const exportMetrics = (client: ClientBase) => {
  /**
   * Export metrics as a CSV file.
   *
   * **Scopes**: `metrics:read`
   *
   * @param query - Query parameters
   * @returns {string}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (query: {
    start_date: string;
    end_date: string;
    timezone?: Timezone;
    interval: TimeInterval;
    organization_id?: string | string[] | null;
    product_id?: string | string[] | null;
    billing_type?: ProductBillingType | ProductBillingType[] | null;
    customer_id?: string | string[] | null;
    metrics?: string[] | null;
  }): Promise<string> => {
    const pathParams = {};
    const queryParams = {
      start_date: query.start_date,
      end_date: query.end_date,
      timezone: query?.timezone || "UTC",
      interval: query.interval,
      organization_id: query?.organization_id,
      product_id: query?.product_id,
      billing_type: query?.billing_type,
      customer_id: query?.customer_id,
      metrics: query?.metrics,
    };
    const request = client.buildRequest(
      "GET",
      "/v1/metrics/export",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<string>(response, "text", {
      422: HTTPValidationError,
    });
  };
};
export const limitsMetrics = (client: ClientBase) => {
  /**
   * Get the interval limits for the metrics endpoint.
   *
   * **Scopes**: `metrics:read`
   *
   * @returns {MetricsLimits}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   */
  return async (): Promise<MetricsLimits> => {
    const pathParams = {};
    const queryParams = {};
    const request = client.buildRequest(
      "GET",
      "/v1/metrics/limits",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<MetricsLimits>(response, "json", {});
  };
};
export const listDashboardsMetrics = (client: ClientBase) => {
  /**
   * List user-defined metric dashboards.
   *
   * **Scopes**: `metrics:read`
   *
   * @param query - Query parameters
   * @returns {MetricDashboardSchema[]}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (query?: {
    organization_id?: string | string[] | null;
  }): Promise<MetricDashboardSchema[]> => {
    const pathParams = {};
    const queryParams = {
      organization_id: query?.organization_id,
    };
    const request = client.buildRequest(
      "GET",
      "/v1/metrics/dashboards",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<MetricDashboardSchema[]>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
export const createDashboardMetrics = (client: ClientBase) => {
  /**
   * Create a user-defined metric dashboard.
   *
   * **Scopes**: `metrics:write`
   *
   * @param body - Request body
   * @returns {MetricDashboardSchema}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (body: MetricDashboardCreate): Promise<MetricDashboardSchema> => {
    const pathParams = {};
    const queryParams = {};
    const request = client.buildRequest(
      "POST",
      "/v1/metrics/dashboards",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<MetricDashboardSchema>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
export const getDashboardMetrics = (client: ClientBase) => {
  /**
   * Get a user-defined metric dashboard by ID.
   *
   * **Scopes**: `metrics:read`
   *
   * @param id - The metric dashboard ID.
   * @returns {MetricDashboardSchema}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string): Promise<MetricDashboardSchema> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "GET",
      "/v1/metrics/dashboards/{id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<MetricDashboardSchema>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
export const deleteDashboardMetrics = (client: ClientBase) => {
  /**
   * Delete a user-defined metric dashboard.
   *
   * **Scopes**: `metrics:write`
   *
   * @param id - The metric dashboard ID.
   * @returns {void}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string): Promise<void> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "DELETE",
      "/v1/metrics/dashboards/{id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<void>(response, "none", {
      422: HTTPValidationError,
    });
  };
};
export const updateDashboardMetrics = (client: ClientBase) => {
  /**
   * Update a user-defined metric dashboard.
   *
   * **Scopes**: `metrics:write`
   *
   * @param id - The metric dashboard ID.
   * @param body - Request body
   * @returns {MetricDashboardSchema}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string, body: MetricDashboardUpdate): Promise<MetricDashboardSchema> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "PATCH",
      "/v1/metrics/dashboards/{id}",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<MetricDashboardSchema>(response, "json", {
      422: HTTPValidationError,
    });
  };
};

export function createMetricsService(client: ClientBase) {
  return {
    get: getMetrics(client),
    export: exportMetrics(client),
    limits: limitsMetrics(client),
    listDashboards: listDashboardsMetrics(client),
    createDashboard: createDashboardMetrics(client),
    getDashboard: getDashboardMetrics(client),
    deleteDashboard: deleteDashboardMetrics(client),
    updateDashboard: updateDashboardMetrics(client),
  };
}

export type Metrics = ReturnType<typeof createMetricsService>;

File: sdk/python/polar/metrics.py
from __future__ import annotations

import typing

from polar.base import (
    AsyncServiceBase,
    SyncServiceBase,
    parse_response_json,
    parse_response_none,
    parse_response_text,
)
from polar.errors import (
    HTTPValidationError,
)
from polar.inputs import (
    MetricDashboardCreate,
    MetricDashboardUpdate,
)
from polar.literals import (
    ProductBillingType,
    TimeInterval,
    Timezone,
)
from polar.outputs import (
    MetricDashboardSchema,
    MetricsLimits,
    MetricsResponse,
)


class MetricsSync(SyncServiceBase):
    def get(
        self,
        *,
        start_date: str,
        end_date: str,
        timezone: Timezone = "UTC",
        interval: TimeInterval,
        organization_id: str | list[str] | None = None,
        product_id: str | list[str] | None = None,
        billing_type: ProductBillingType
        | list[ProductBillingType]
        | None = None,
        customer_id: str | list[str] | None = None,
        metrics: list[str] | None = None,
    ) -> MetricsResponse:
        """
        Get metrics about your orders and subscriptions.

        Currency values are output in cents.

        **Scopes**: `metrics:read`

        Args:
            start_date: Start date.
            end_date: End date.
            timezone: Timezone to use for the timestamps. Default is UTC.
            interval: Interval between two timestamps.
            organization_id: Filter by organization ID.
            product_id: Filter by product ID.
            billing_type: Filter by billing type. `recurring` will filter data corresponding to subscriptions creations or renewals. `one_time` will filter data corresponding to one-time purchases.
            customer_id: Filter by customer ID.
            metrics: List of metric slugs to focus on. When provided, only the queries needed for these metrics will be executed, improving performance. If not provided, all metrics are returned.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/metrics/",
            path_params={},
            query_params={
                "start_date": start_date,
                "end_date": end_date,
                "timezone": timezone,
                "interval": interval,
                "organization_id": organization_id,
                "product_id": product_id,
                "billing_type": billing_type,
                "customer_id": customer_id,
                "metrics": metrics,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, MetricsResponse, method_errors)

    def export(
        self,
        *,
        start_date: str,
        end_date: str,
        timezone: Timezone = "UTC",
        interval: TimeInterval,
        organization_id: str | list[str] | None = None,
        product_id: str | list[str] | None = None,
        billing_type: ProductBillingType
        | list[ProductBillingType]
        | None = None,
        customer_id: str | list[str] | None = None,
        metrics: list[str] | None = None,
    ) -> str:
        """
        Export metrics as a CSV file.

        **Scopes**: `metrics:read`

        Args:
            start_date: Start date.
            end_date: End date.
            timezone: Timezone to use for the timestamps. Default is UTC.
            interval: Interval between two timestamps.
            organization_id: Filter by organization ID.
            product_id: Filter by product ID.
            billing_type: Filter by billing type. `recurring` will filter data corresponding to subscriptions creations or renewals. `one_time` will filter data corresponding to one-time purchases.
            customer_id: Filter by customer ID.
            metrics: List of metric slugs to include in the export. If not provided, all metrics are exported.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/metrics/export",
            path_params={},
            query_params={
                "start_date": start_date,
                "end_date": end_date,
                "timezone": timezone,
                "interval": interval,
                "organization_id": organization_id,
                "product_id": product_id,
                "billing_type": billing_type,
                "customer_id": customer_id,
                "metrics": metrics,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_text(response, method_errors)

    def limits(
        self,
    ) -> MetricsLimits:
        """
        Get the interval limits for the metrics endpoint.

        **Scopes**: `metrics:read`

        Args:

        Raises:
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/metrics/limits",
            path_params={},
            query_params={},
        )
        response = self.client.send_request(request)
        return parse_response_json(response, MetricsLimits)

    def list_dashboards(
        self,
        *,
        organization_id: str | list[str] | None = None,
    ) -> list[MetricDashboardSchema]:
        """
        List user-defined metric dashboards.

        **Scopes**: `metrics:read`

        Args:
            organization_id: Filter by organization ID.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/metrics/dashboards",
            path_params={},
            query_params={
                "organization_id": organization_id,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(
            response, list[MetricDashboardSchema], method_errors
        )

    def create_dashboard(
        self,
        **kwargs: typing.Unpack[MetricDashboardCreate],
    ) -> MetricDashboardSchema:
        """
        Create a user-defined metric dashboard.

        **Scopes**: `metrics:write`

        Args:
            **kwargs: Request body parameters

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/metrics/dashboards",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, MetricDashboardSchema, method_errors)

    def get_dashboard(
        self,
        id: str,
    ) -> MetricDashboardSchema:
        """
        Get a user-defined metric dashboard by ID.

        **Scopes**: `metrics:read`

        Args:
            id: The metric dashboard ID.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/metrics/dashboards/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, MetricDashboardSchema, method_errors)

    def delete_dashboard(
        self,
        id: str,
    ) -> None:
        """
        Delete a user-defined metric dashboard.

        **Scopes**: `metrics:write`

        Args:
            id: The metric dashboard ID.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="DELETE",
            url="/v1/metrics/dashboards/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_none(response, method_errors)

    def update_dashboard(
        self,
        id: str,
        **kwargs: typing.Unpack[MetricDashboardUpdate],
    ) -> MetricDashboardSchema:
        """
        Update a user-defined metric dashboard.

        **Scopes**: `metrics:write`

        Args:
            id: The metric dashboard ID.
            **kwargs: Request body parameters

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/metrics/dashboards/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, MetricDashboardSchema, method_errors)


class MetricsAsync(AsyncServiceBase):
    async def get(
        self,
        *,
        start_date: str,
        end_date: str,
        timezone: Timezone = "UTC",
        interval: TimeInterval,
        organization_id: str | list[str] | None = None,
        product_id: str | list[str] | None = None,
        billing_type: ProductBillingType
        | list[ProductBillingType]
        | None = None,
        customer_id: str | list[str] | None = None,
        metrics: list[str] | None = None,
    ) -> MetricsResponse:
        """
        Get metrics about your orders and subscriptions.

        Currency values are output in cents.

        **Scopes**: `metrics:read`

        Args:
            start_date: Start date.
            end_date: End date.
            timezone: Timezone to use for the timestamps. Default is UTC.
            interval: Interval between two timestamps.
            organization_id: Filter by organization ID.
            product_id: Filter by product ID.
            billing_type: Filter by billing type. `recurring` will filter data corresponding to subscriptions creations or renewals. `one_time` will filter data corresponding to one-time purchases.
            customer_id: Filter by customer ID.
            metrics: List of metric slugs to focus on. When provided, only the queries needed for these metrics will be executed, improving performance. If not provided, all metrics are returned.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/metrics/",
            path_params={},
            query_params={
                "start_date": start_date,
                "end_date": end_date,
                "timezone": timezone,
                "interval": interval,
                "organization_id": organization_id,
                "product_id": product_id,
                "billing_type": billing_type,
                "customer_id": customer_id,
                "metrics": metrics,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, MetricsResponse, method_errors)

    async def export(
        self,
        *,
        start_date: str,
        end_date: str,
        timezone: Timezone = "UTC",
        interval: TimeInterval,
        organization_id: str | list[str] | None = None,
        product_id: str | list[str] | None = None,
        billing_type: ProductBillingType
        | list[ProductBillingType]
        | None = None,
        customer_id: str | list[str] | None = None,
        metrics: list[str] | None = None,
    ) -> str:
        """
        Export metrics as a CSV file.

        **Scopes**: `metrics:read`

        Args:
            start_date: Start date.
            end_date: End date.
            timezone: Timezone to use for the timestamps. Default is UTC.
            interval: Interval between two timestamps.
            organization_id: Filter by organization ID.
            product_id: Filter by product ID.
            billing_type: Filter by billing type. `recurring` will filter data corresponding to subscriptions creations or renewals. `one_time` will filter data corresponding to one-time purchases.
            customer_id: Filter by customer ID.
            metrics: List of metric slugs to include in the export. If not provided, all metrics are exported.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/metrics/export",
            path_params={},
            query_params={
                "start_date": start_date,
                "end_date": end_date,
                "timezone": timezone,
                "interval": interval,
                "organization_id": organization_id,
                "product_id": product_id,
                "billing_type": billing_type,
                "customer_id": customer_id,
                "metrics": metrics,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_text(response, method_errors)

    async def limits(
        self,
    ) -> MetricsLimits:
        """
        Get the interval limits for the metrics endpoint.

        **Scopes**: `metrics:read`

        Args:

        Raises:
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/metrics/limits",
            path_params={},
            query_params={},
        )
        response = await self.client.send_request(request)
        return parse_response_json(response, MetricsLimits)

    async def list_dashboards(
        self,
        *,
        organization_id: str | list[str] | None = None,
    ) -> list[MetricDashboardSchema]:
        """
        List user-defined metric dashboards.

        **Scopes**: `metrics:read`

        Args:
            organization_id: Filter by organization ID.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/metrics/dashboards",
            path_params={},
            query_params={
                "organization_id": organization_id,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(
            response, list[MetricDashboardSchema], method_errors
        )

    async def create_dashboard(
        self,
        **kwargs: typing.Unpack[MetricDashboardCreate],
    ) -> MetricDashboardSchema:
        """
        Create a user-defined metric dashboard.

        **Scopes**: `metrics:write`

        Args:
            **kwargs: Request body parameters

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/metrics/dashboards",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, MetricDashboardSchema, method_errors)

    async def get_dashboard(
        self,
        id: str,
    ) -> MetricDashboardSchema:
        """
        Get a user-defined metric dashboard by ID.

        **Scopes**: `metrics:read`

        Args:
            id: The metric dashboard ID.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/metrics/dashboards/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, MetricDashboardSchema, method_errors)

    async def delete_dashboard(
        self,
        id: str,
    ) -> None:
        """
        Delete a user-defined metric dashboard.

        **Scopes**: `metrics:write`

        Args:
            id: The metric dashboard ID.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="DELETE",
            url="/v1/metrics/dashboards/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_none(response, method_errors)

    async def update_dashboard(
        self,
        id: str,
        **kwargs: typing.Unpack[MetricDashboardUpdate],
    ) -> MetricDashboardSchema:
        """
        Update a user-defined metric dashboard.

        **Scopes**: `metrics:write`

        Args:
            id: The metric dashboard ID.
            **kwargs: Request body parameters

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarRateLimitError: Raised when the rate limit is exceeded.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/metrics/dashboards/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, MetricDashboardSchema, method_errors)
